src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -811,6 +827,26 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded, Immu
                 {
                     throw ExceptionUtilities.Unreachable;
                 }
+
+                // Local functions
+                static SourceGeneratedDocumentState? FindExistingGeneratedDocumentState(
+                    ImmutableArray<SourceGeneratedDocumentState> states,
+                    ISourceGenerator generator,
+                    string hintName)
+                {
+                    foreach (var state in states)
```

Comment:
📝 This is an O(n²) check (where _n_ is the number of source generated documents) since we loop through all states for each new document.

Reply:
seems fine.  if it's a problem, would be trivial to switch to a faster data structure in the future.